### ISSUE TITLE
[Proposal] Verify that Input Tensor is a probability in classification metrics

### ISSUE THREAD

**Issue by maronuu** [pytorch/torcheval#115]
### 🚀 The feature

Add the verification to a series of functions like [_update_input_check](https://github.com/pytorch/torcheval/blob/e80aa1f584ee8e59c0031a3e8e74bc3e12be4b5b/torcheval/metrics/functional/classification/accuracy.py#L360) in clasification metrics that `input` tensor is a probability, i.e., all elements satisfy `0 <= x <= 1`.

### Motivation, pitch

Hi torcheval community!
This is an enhancement suggestions I came up with while using torcheval for an image classification project.

Here is a [binary_accuracy(classification)](https://pytorch.org/torcheval/stable/generated/torcheval.metrics.BinaryAccuracy.html#torcheval.metrics.BinaryAccuracy) documentation. Judging from the default argument `threshold = 0.5`, torcheval assumes that `input` is a probability. However, there is no statement or intent in the documentation that input must be a probability. There is no validation in `_input_check` functions, too.

There is a problem when the model does not include the Sigmoid or Softmax layer. This can often happen when they use torch Loss classes like `BCEWithLogitLoss` or `CrossEntropyLoss`, which include sigmoid/softmax functions in themselves.

Here is a code including a bug.
```python
class Trainer:
    def __init__(self, model: nn.Module):
        self._loss = nn.BCEWithLogitLoss() # performs `sigmoid` in itself
        self._model = model # model does not end with `sigmoid` layer.
...
    def train(self, batch):
        img, correct_label = batch
        out = self._model(img) 
        loss = self._loss(out, correct_label)
        
        # out is not 'probability', but `torcheval.binary_accuracy` function can be run.
        acc = binary_accuracy(out, correct_label)
```

IMO, this wrong code can be written in a natural manner, because the scope of responsibility of `binary_accuracy` function is unclear.
Clarifying the responsibility of classification metric funcions in torcheval will remove these bug cases.

#### Approach 1: Limit use cases by including a cautionary note in the Document
Adding note that "input is assumed to be a probability" to doc.

Pros: easy, no effect for existing code.
Cons: user must be responsible that `input` is a probability.
#### Approach 2: Adding some `assert` codes
Adding a verification `0 <= input <= 1` to `_input_check` family.

Pros: Confusion written in the `Background` section can be avoided.
Cons: This needs to be implemented. may have effects on existing codes.

**If this proposal is deemed to be viable, I am willing to contribute it!.**

Thank you.

**Comment by bobakfb** [comment]
Hi @maronuu, sorry for the slow reply during the Holidays, and thanks for bringing this issue!

Hmm... well my immediate opinion is that I don't believe we require the inputs to be probabilities, even though we do expect that will be how most people use these metrics. I'm not really sure we'd be helping anyone if we require the inputs be probabilities. That said, I'll leave this open for the time being to let others give their opinions.
